Run jobs on demand

bvallelunga opened the pull request:
We currently use [Agenda](https://github.com/agenda/agenda) paired with [AgendaDash](https://github.com/agenda/agendash). One of the features AgendaDash provides is an ability to run jobs at any time, not just when they are failed state. We consider this a core feature that any replacement should have. 

We are moving to Bull/Arena and would love this feature to be included.

bvallelunga commented:
@skeggse @ttacon would love your review.

skeggse commented on src/server/views/api/jobRetry.js:
```diff
@@ -10,7 +10,12 @@ async function handler(req, res) {
 
   try {
-    await job.retry();
+    if(await job.getState() === "failed") {
```
`getState` isn't a function provided by bee's `Job` interface.

skeggse commented on src/server/views/api/jobRetry.js:
```diff
@@ -10,7 +10,12 @@ async function handler(req, res) {
+    if(await job.getState() === "failed") {
+      await job.retry();
+    } else {
+      await queue.add(job.name, job.data)
```
`Queue#add` isn't provided by bee's `Queue` interface.

bvallelunga commented:
> This isn't functionality that's supported by bee. I can't speak to whether it's supported by bull. If it is, happy to approve with a guard that disables the additional functionality.
> 
> Also, please try to match the existing code style when submitting PRs. We should probably adopt `prettier` in this project just to make this a non-issue.

Thank @skeggse for the review! I just added functionality so that it works with both Bee and Bull. I think I updated the code style to match, if I missed anything let me know.

skeggse commented on src/server/queue/index.js:
```diff
@@ -85,10 +85,16 @@ class Queues {
+        }
+      ];
+
+      if(data.name) args.unshift(data.name)
```
```suggestion
      if (data.name) args.unshift(data.name)
```

skeggse commented on src/server/queue/index.js:
```diff
@@ -85,10 +85,16 @@ class Queues {
-        removeOnComplete: false,
-        removeOnFail: false
-      });
+      let args = [
```
```suggestion
      const args = [
```

skeggse commented on src/server/queue/index.js:
```diff
@@ -85,10 +85,16 @@ class Queues {
+      ];
+
+      if(data.name) args.unshift(data.name)
+      return queue.add(...args);
```
bee-queue currently supports Node 4, so we can't use this syntax.

skeggse commented on src/server/views/api/jobRetry.js:
```diff
@@ -10,7 +10,19 @@ async function handler(req, res) {
 
   try {
-    await job.retry();
+    let jobState;
+    if (queue.IS_BEE) {
+      jobState = job.status;
+    } else {
+      jobState = await job.getState();
+    }
```
truly just a suggestion - not a necessary change
```suggestion
    const jobState = queue.IS_BEE ? job.status : await job.getState();
```

skeggse commented on src/server/views/api/jobRetry.js:
```diff
@@ -10,7 +10,19 @@ async function handler(req, res) {
+      jobState = await job.getState();
+    }
+
+    if(jobState === "failed" && typeof job.retry === "function") {
```
```suggestion
    if (jobState === 'failed' && typeof job.retry === 'function') {
```

skeggse commented on src/server/views/partials/dashboard/jobDetails.hbs:
```diff
@@ -7,11 +7,9 @@
-    Retry
-  </button>
-{{/eq}}
+<button class="btn btn-success js-retry-job" data-queue-host="{{ queueHost }}" data-queue-name="{{ queueName }}" data-job-id="{{ this.id }}">
```
This button will be visible even if it's not available due to the job residing on a bee-queue server. I'd like to preserve the behavior that the button only appears for failed jobs in bee-queue.

bvallelunga replied:
I am not that familiar with how handlebars and locals work, is there an easy way in the view to check for this. I am thinking the if statement should be `!queue.IS_BEE || queue.state == "failed"`

skeggse replied:
The recommended practice is that you introduce a new local like `showRetryButton` that computes the check in js and then you can just do `{{#if showRetryButton}}` without encoding complex logic in the template itself.

bvallelunga replied:
Scratch that, I figured it out.

bvallelunga replied:
> The recommended practice is that you introduce a new local like `showRetryButton` that computes the check in js and then you can just do `{{#if showRetryButton}}` without encoding complex logic in the template itself.

Just saw this. That makes sense, let me know if the way I did it works for you.

skeggse replied:
Hey, no worries! I would strongly prefer a DRYer approach where we don't duplicate the markup thrice - the `showRetryButton` approach would achieve that.

bvallelunga replied:
Awesome! Just made the changes.

skeggse commented on src/server/views/api/jobRetry.js:
```diff
@@ -10,7 +10,19 @@ async function handler(req, res) {
+    if(jobState === "failed" && typeof job.retry === "function") {
+      await job.retry();
+    } else {
+      await Queues.set(queue, job);
```
Not sure how Bull's `name` feature works, but does it belong as a parameter here too so it can get passed to `queue.add`?

bvallelunga replied:
It is already in the bull job object. In `Queues.set` we extract it before calling `queue.add`

skeggse commented on src/server/views/dashboard/jobDetails.js:
```diff
@@ -19,14 +19,9 @@ async function handler(req, res) {
-    job.logs = (logs.logs || "No Logs");
-  }
+  const jobState = queue.IS_BEE ? job.status : await job.getState();
+  job.showRetryButton = !queue.IS_BEE || jobState == 'failed';
+  job.retryButtonText = jobState == 'failed' ? 'Retry' : 'Trigger';
```
strict equality for unsurprising behavior
```suggestion
  job.showRetryButton = !queue.IS_BEE || jobState === 'failed';
  job.retryButtonText = jobState === 'failed' ? 'Retry' : 'Trigger';
```

skeggse commented on src/server/views/dashboard/queueJobsByState.js:
```diff
@@ -103,6 +103,13 @@ async function _html(req, res) {
 
+  await jobs.map(async (job) => {
+    const jobState = queue.IS_BEE ? job.status : await job.getState();
+    job.showRetryButton = !queue.IS_BEE || jobState == 'failed';
+    job.retryButtonText = jobState == 'failed' ? 'Retry' : 'Trigger';
```
```suggestion
    job.showRetryButton = !queue.IS_BEE || jobState === 'failed';
    job.retryButtonText = jobState === 'failed' ? 'Retry' : 'Trigger';
```

skeggse commented on src/server/views/dashboard/queueJobsByState.js:
```diff
@@ -103,6 +103,13 @@ async function _html(req, res) {
     })
   }
 
+  await jobs.map(async (job) => {
```
This will result in `await`ing an `Array`, which won't actually wait for the promises within the array to complete. This can result in unhandled rejections, and likely introduces a race condition or other incorrect behavior resulting in the button not appearing when it should.

skeggse commented on src/server/views/dashboard/queueJobsByState.js:
```diff
@@ -103,6 +103,13 @@ async function _html(req, res) {
+    const jobState = queue.IS_BEE ? job.status : await job.getState();
+    job.showRetryButton = !queue.IS_BEE || jobState == 'failed';
+    job.retryButtonText = jobState == 'failed' ? 'Retry' : 'Trigger';
+    return job
```
Not used
```suggestion
```

skeggse commented on src/server/views/dashboard/queueJobsByState.js:
```diff
@@ -103,6 +103,13 @@ async function _html(req, res) {
+    job.showRetryButton = !queue.IS_BEE || jobState == 'failed';
+    job.retryButtonText = jobState == 'failed' ? 'Retry' : 'Trigger';
+    return job
+  })
```
```suggestion
  });
```

skeggse commented on src/server/views/partials/dashboard/jobDetails.hbs:
```diff
@@ -7,11 +7,12 @@
+{{#if showRetryButton}}
   <button class="btn btn-success js-retry-job" data-queue-host="{{ queueHost }}" data-queue-name="{{ queueName }}" data-job-id="{{ this.id }}">
-    Retry
+    {{ retryButtonText }}
```
style consistency
```suggestion
    {{retryButtonText}}
```

bvallelunga commented:
> Almost there! I'm mostly concerned about one logical error.

Awesome!! Just put up an update!

skeggse commented on src/server/views/api/jobRetry.js:
```diff
@@ -10,7 +10,14 @@ async function handler(req, res) {
-    await job.retry();
+    const jobState = queue.IS_BEE ? job.status : await job.getState();
+
+    if(jobState === 'failed' && typeof job.retry === 'function') {
```
```suggestion
    if (jobState === 'failed' && typeof job.retry === 'function') {
```

skeggse commented on src/server/views/partials/dashboard/jobDetails.hbs:
```diff
@@ -7,11 +7,12 @@
+{{#if showRetryButton}}
   <button class="btn btn-success js-retry-job" data-queue-host="{{ queueHost }}" data-queue-name="{{ queueName }}" data-job-id="{{ this.id }}">
-    Retry
+    {{retryButtonText}}
```
```suggestion
    {{ retryButtonText }}
```

bvallelunga commented:
@skeggse thanks for merging! What is your guys typical timeline to release to NPM. I would love to make the switch from Agenda to Bull this week if it's possible.

skeggse commented:
Landed in `v2.8.0`.

bvallelunga commented:
Amazing! Super exicting

bvallelunga commented:
@skeggse I just put up another PR that fixes an issue with named processes (something the Bull community has struggled with for a bit).
